db/src/main/scala/slick/SlickDb.scala

```
@@ -20,16 +23,35 @@ object SlickDb {
     profile: JdbcProfile,
     dialect: Dialect[?],
   ): F[SlickDb] = {
-    // Execute migrations each time SlickDb is instantiated
-    def runMigration(db: Database): F[Unit] =
-      Async[F].fromFuture(Sync[F].delay(db.run(migrations.all(dialect)()))).handleErrorWith {
-        // Since slick-migration-api has no information about which migrations have been applied,
-        // re-applying migrations may throw exception about the existence of an entity in the database
-        // SQLState 42P07 means `ERROR: relation "xxx" already exists`
-        case e: PSQLException if e.getSQLState == "42P07" => ().pure
-        case error                                        => Async[F].raiseError(error)
-      }
-
-    runMigration(db).as(new SlickDb(db, profile))
+    def runMigrations(db: Database): F[SlickDb] = {
+      implicit val slickDb: SlickDb = new SlickDb(db, profile)
+      val key                       = "dbVersion"
+
+      def applyAllNewerThen(version: Int, api: SlickApi[F]): F[Unit] = migrations
+        .all(dialect)
+        .drop(version)
+        .toList
+        .traverse_ { case (idx, migrations) =>
+          import profile.api.*
+          Async[F].executionContext.flatMap { implicit ec =>
+            val actions = DBIO
+              .sequence(migrations.migrations.map(m => m()))
+              .flatMap(_ => api.queries.putConfig(key, idx.toString))
+            actions.transactionally.run
+          }
+        }
+
+      def run: F[Unit] = (for {
+        api          <- SlickApi[F](slickDb)
+        dbVersionOpt <- api.queries.getConfig(key).run.recover { case _ => "0".some }
+        version      <- Try(dbVersionOpt.getOrElse("0").toInt)
+                          .adaptErr(_ => new RuntimeException(s"Error reading $key from config table"))
```

Lets use `sdk.error.FatalError` here
`.adaptErr(FatalError(s"Error reading $key from config table", _))`

we should not swallow errors but chain them.